crates/driver/src/tests/cases/protocol_fees.rs

```diff
@@ -187,3 +188,130 @@ async fn volume_protocol_fee_sell_order() {
 
     protocol_fee_test_case(test_case).await;
 }
+
+#[tokio::test]
+#[ignore]
+async fn price_improvement_fee_buy_out_market_order() {
+    let quote_sell_amount = 50000000000000000000u128;
+    let quote_buy_amount = 45000000000000000000u128;
+    let fee_policy = FeePolicy::PriceImprovement {
+        factor: 0.5,
+        // high enough so we don't get capped by volume fee
+        max_volume_factor: 1.0,
+        quote: PriceImprovementQuote {
+            sell_amount: quote_sell_amount.into(),
+            buy_amount: quote_buy_amount.into(),
+            fee: 1000000000000000000u128.into(),
+        },
+    };
+    let test_case = TestCase {
+        order_side: order::Side::Buy,
+        fee_policy,
+        order_sell_amount: 50000000000000000000u128.into(),
+        solver_fee: Some(10000000000000000000u128.into()),
+        quote_sell_amount: quote_sell_amount.into(),
+        quote_buy_amount: quote_buy_amount.into(),
+        executed: quote_buy_amount.into(),
+        // sell amount + quote fee * factor
+        executed_sell_amount: 50500000000000000000u128.into(),
+        // executed buy amount should match quote buy amount
+        executed_buy_amount: quote_buy_amount.into(),
+    };
+
+    protocol_fee_test_case(test_case).await;
+}
+
+#[tokio::test]
+#[ignore]
+async fn price_improvement_fee_sell_out_market_order() {
+    let quote_sell_amount = 50000000000000000000u128;
+    let quote_buy_amount = 45000000000000000000u128;
+    let fee_policy = FeePolicy::PriceImprovement {
+        factor: 0.5,
+        // high enough so we don't get capped by volume fee
+        max_volume_factor: 1.0,
+        quote: PriceImprovementQuote {
+            sell_amount: quote_sell_amount.into(),
+            buy_amount: quote_buy_amount.into(),
+            fee: 1000000000000000000u128.into(),
+        },
+    };
+    let order_sell_amount = 50000000000000000000u128;
+    let test_case = TestCase {
+        order_side: order::Side::Sell,
+        fee_policy,
+        order_sell_amount: order_sell_amount.into(),
+        solver_fee: Some(10000000000000000000u128.into()),
```

Comment:
nit: I think it would be clearer to call this `network_fee` in this context.

I do see now that the price improvement comes from the network fee being different than the quoted fee. This is very clever but not something I'd expect someone maintaining this code in 3 month to just guess without context, so please let's help your future colleague to understand how this test work more intuitively.